### ISSUE TITLE
How to parse the request data using Form and still log the Request body?

### ISSUE THREAD

**Issue by gor-bit** [tiangolo/fastapi#3857]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import Request, FastAPI, Form

app = FastAPI()

@app.post("/post_name")
async def parse_data(request: Request, first_name: str = Form(...), last_name: str = Form(...)):
    data_body = await request.body()
    logging.info(data_body)  
    logging.info(first_name)
    logging.info(last_name)
    return first_name + last_name
```

### Description

I have an endpoint `/post_name` for users to send their `first_name` and `last_name` as form data.

This can be simulated using a curl command:
```
curl -d first_name=Sabrina -d last_name=Alami -X POST website.com/post_name
```

However, occasionally I may receive people who accidentally send their data using `second_name` instead of `last_name`. Or they may include another parameter such a `middle_name` due to the common tradition of my country which results in an error.

For requests which are successful, I am able to log the `first_name` and `last_name` fields easily. But I want to be able to log the body of all requests, including those which result in an error. Hence I want to log the body of **all** requests received so I don't lose this information.

The above code snippet should do this, but gives me the following error whenever I receive a request:
```
RuntimeError: Stream consumed
```

What am I doing wrong when trying to access `request.body()`?

Thank you.

### Operating System

Linux

### FastAPI Version

0.68.1

### Python Version

Python 3.8.10

**Comment by frankie567** [comment]
Use the `.form` method instead of `.body`, it'll directly give you a parsed version of the form payload. Since internally FastAPI already called it to parse `first_name` and `last_name`, you don't consume the stream twice, the result is already cached:

```py
from fastapi import Request, FastAPI, Form

app = FastAPI()

@app.post("/post_name")
async def parse_data(request: Request, first_name: str = Form(...), last_name: str = Form(...)):
    data_body = await request.form()
    logging.info(data_body)  
    logging.info(first_name)
    logging.info(last_name)
    return first_name + last_name
```

Ref: https://www.starlette.io/requests/#body

**Comment by gor-bit** [comment]
@frankie567 This solution works if a user passes in the correct parameters e.g. `first_name` and `last_name`, along with any extra parameters not needed by the API.

However, if the user fails to pass in all of the correct parameters, e.g. they pass in `first_name` and `second_name`, only a `422` response is logged. The form payload is not logged in this case.

The following curl command demonstrates this behavior with your provided code:
```
curl -d first_name=Sabrina -d second_name=Alami -X POST website.com/post_name
```

Is there a way to still capture the form payload even if I get a `422` response?

**Comment by frankie567** [comment]
Yes, that's how FastAPI works: if you define it as a required parameter in your endpoint, FastAPI will trigger its validation mechanism.

If you need something completely dynamic, you'll have to remove those parameters and perform the validation yourself:

```py
from fastapi import Request, FastAPI, Form, HTTPException

app = FastAPI()

@app.post("/post_name")
async def parse_data(request: Request):
    data_body = await request.form()
    logging.info(data_body)  
    if "first_name" not in data_body or "last_name" not in data_body:
        raise HTTPException(422, "Missing fields")
    return first_name + last_name
```

**Comment by gor-bit** [comment]
That is very helpful, thank you @frankie567 . 

For completeness, this is the final code which meets my requirements:

```python
from fastapi import Request, FastAPI, Form, HTTPException
import logging

app = FastAPI()

@app.post("/post_name")
async def parse_data(request: Request):
    data_body = await request.form()
    logging.info(data_body)  
    if "first_name" not in data_body or "last_name" not in data_body:
        raise HTTPException(422, "Missing fields")
    first_name = data_body["first_name"]
    last_name = data_body["last_name"]
    return first_name + last_name
```
